### DISCUSSION TITLE
Someone working on updating documentation with the new version of vEOS ?

### DISCUSSION THREAD

**Discussion by htam-net** [ipspace/netlab#408] [Ideas]
With the new versions of vEOS that require the use of Aboot iso file as cdrom, is someone working on documenting the use of these new images (like 4.28) ?

**Comment by ipspace** [comment]
Not yet. Help would be appreciated ;)

**Reply by htam-net** [reply]
Just to share : I was able to start the new image with some tweaking of the XML profile in libvirt/virt-manager. I don't know how to translate that in Vagrant to create such Box... It seems that the experimental version 2 of Box format may be suited for this usage, if it is able to specify the type of disk for one disk image (cdrom not harddrive).

**Reply by ipspace** [reply]
Are you saying the new EOS image needs a CDROM and a harddrive to boot? That would be a major PITA.

**Reply by htam-net** [reply]
Yes : Here is the XML part of libvirt definition for the vEOS VM (the device/disk order and name are very strict - from my experience) : 
```
    <disk type='file' device='disk'>
      <driver name='qemu' type='vmdk'/>
      <source file='./arista/vEOS64-lab-4.28.2F.vmdk'/>
      <target dev='hda' bus='ide'/>
      <alias name='ide0-0-0'/>
      <address type='drive' controller='0' bus='0' target='0' unit='0'/>
    </disk>
    <disk type='file' device='cdrom'>
      <driver name='qemu' type='raw'/>
      <source file='./arista/Aboot-veos-8.0.1.iso'/>
      <target dev='hdc' bus='ide'/>
      <readonly/>
      <alias name='ide0-1-0'/>
      <address type='drive' controller='0' bus='1' target='0' unit='0'/>
    </disk>
```
Based from this Arista [documentation](https://www.arista.com/en/cg-veos-router/veos-router-using-the-veos-router-on-kvm-and-esxi)
Also, from this [cool page](https://codingpackets.com/blog/arista-veos-vagrant-libvirt-box-install/) that inspire me, the Arista configuration syntax has changed : ZTP has to be disable, SSH Key syntax has changed, ...

**Comment by ipspace** [comment]
Well, you had me worried for a while, but the sky is not falling (yet).

* I downloaded `vEOS-lab-4.28.2F.qcow2` from Arista's web site
* I moved it into an empty directory and started `netlab libvirt package eos vEOS-lab-4.28.2F.qcow2`
* The VM booted as expected
* I followed the instructions in `netlab libvirt config eos` and got a working vEOS box
* I changed the ~/.netlab.yml settings to use the new box and started a lab that I wanted to work on anyway.

Turns out that you still don't need 4GB to get it up and running -- I have three instances consuming 2GB each running VRF Lite lab at the moment.

**Reply by htam-net** [reply]
Thank you. OK it has to do with my setup . For documentation, the Aboot part is booting, but not the swi image : 

```
$ netlab libvirt package eos vEOS-lab-4.28.2F.qcow2  

=================
     WARNING
=================
This is an experimental script that does its best to build a Vagrant box for libvirt
provider out of a VM disk. It might die a horrible death and leave all sorts of garbage
behind that you'll have to clean up by hand (for example, libvirt 'vm_box' virtual machine).

It also assumes that it can wreak havoc in the current directory (although it will do its
best not to damage the original virtual disk).


Do you want to continue [Y/n]: Y
error: failed to get domain 'vm_box'

error: failed to get domain 'vm_box'

Creating a copy of vEOS-lab-4.28.2F.qcow2

====================
Starting the VM
====================
We'll start the VM from the newly-created virtual disk. When the
VM starts, execute 'netlab libvirt config eos' in another
window and follow the instructions.
====================


WARNING  OS name 'linux' is deprecated, using 'generic' instead. This alias will be removed in the future.

Starting install...
Running text console command: virsh --connect qemu:///system console vm_box
Connected to domain 'vm_box'
Escape character is ^] (Ctrl + ])
[   15.566712] ==== Dumping storage debug info for flash ()
[   15.567877] Failed to find block device for flash


Aboot 8.0.1
Segmentation fault
Timed out waiting for /mnt/flash
cp: can't stat '/mnt/flash.conf': No such file or directory
Segmentation fault
Segmentation fault

Segmentation fault
Welcome to Aboot.
Aboot# ^M

Aboot# 
Aboot# [   77.044487] init[1]: segfault at ffdcf958 ip 00000000080482ce sp 00000000ffdcf958 error 25 in init[8048000+cb000]
[   77.046333] Code: Bad RIP value.
[   77.046988] Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000000b
[   77.046988] 
[   77.049793] Kernel Offset: disabled
[   77.050495] panic_event_handler: unable to lock eeprom
[   77.051489] Rebooting in 60 seconds..

ISOLINUX 4.05 0x5be438e0  Copyright (C) 1994-2011 H. Peter Anvin et al
Loading linux......
Loading initrd...ready.
[   15.553652] ==== Dumping storage debug info for flash ()
[   15.555352] Failed to find block device for flash
[   16.587436] Running e2fsck on: /mnt/flash
[   16.606898] e2fsck on /mnt/flash took 0s


Aboot 8.0.1
Segmentation fault
Timed out waiting for /mnt/flash
cp: can't stat '/mnt/flash.conf': No such file or directory
Segmentation fault

Segmentation fault
Segmentation fault
Type "fullrecover" and press Enter to revert /mnt/flash to factory default
state, or just press Enter to reboot: 

```

**Reply by ipspace** [reply]
Hope you're working on a throwaway server/VM ;)

If that's the case, and if you're running Ubuntu, try **netlab install ubuntu libvirt** (hoping there's something obvious missing in your setup). Would love to know what's wrong (or why it's working for me).

**Reply by htam-net** [reply]
for the record I'm using an AMD Ryzen 5 5600X cpu.
The latest vEOS-lab image I'm able to boot with libvirt is vEOS-lab-combined-4.20.15M.vmdk . I have tried 32bits and 64bits version of 4.28.2F and 4.27.6M in vmdk and qcow2 formats, with and without Aboot iso image, without success : either /mnt/flash "empty" or segmentation fault.

**Reply by ipspace** [reply]
Interesting, so it looks like vEOS doesn't work with AMD CPU. @Kircheneer had a similar problem with vSRX. Try copying

```
      {% if "amd" in defaults.processor|lower %}
      domain.cpu_mode = "custom"
      {% endif %}
```

... from `netsim/templates/provider/libvirt/vsrx-domain.j2` to `netsim/templates/provider/libvirt/eos-domain.js`.

**Reply by ipspace** [reply]
... forgot you're still building the box. You would have to add a corresponding setting to `devices.eos.libvirt.create` system default (`netsim/topology-defaults.yml`) (probably change `--cpu host` to `--cpu custom` or some such thing).

**Comment by petercrocker** [comment]
For the record, I run arista 4.18 (or so?) on an AMD Threadripper 39xx CPU just fine, and also vSRX on that CPU, no problems, no special configs. Let me know if it's helpful to dig out examples.

**Reply by ipspace** [reply]
4.18 is a bit old. We're talking about 4.28 ;)

**Comment by ipspace** [comment]
Stepping back a bit (now that we're at the "this is box-specific or AMD-specific issue")

* I'm assuming you enabled all the necessary VT bits in BIOS
* What Linux distro are you running?
* Are you running Linux on a bare-metal server or within a VM?
* If you're running Linux within a VM, what is the host operating system, and what virtualization solution are you using?

Not that I'd have any specific suggestions based on what your answers might be (as I never had a machine with an AMD CPU), but at least we'd have a well-defined data point that we could add to the documentation.

**Reply by htam-net** [reply]
Sorry for the delay : 
Yes, everything VT is enabled (SVM mode on AMD)
I'm running Debian 11.4 (and now .5) on bare-metal PC (with a lot of RAM).

**Reply by ipspace** [reply]
Here's another thing you could try (please note I'm throwing pasta at the wall to see if it sticks)

It could be that there's something in your Debian setup that annoys the EOS VM. No idea what it could be, but I've experienced a few crazy issues in the last few days, so I'm slowly starting to believe in magic.

You could create a Ubuntu VM using the installation recipe (https://netsim-tools.readthedocs.io/en/latest/install/ubuntu.html) and then build the Arista box inside that VM (assuming nested KVM works as expected). If that doesn't work, then it's probably hardware related (I know people running EOS boxes in similar environments), otherwise it could be the Debian setup.

**Reply by htam-net** [reply]
I built an Ubuntu 22.04.1 LTS VM on my Debian PC. Inside the VM, I was able to boot vEOS64-lab-4.28.2F.qcow2 using this page : https://codingpackets.com/blog/arista-veos-vagrant-libvirt-box-install/ 
FYI, the `netlab libvirt package eos vEOS64-lab-4.28.2F.qcow2` gives an error for deprecated option :

```
We'll start the VM from the newly-created virtual disk. When the
VM starts, execute 'netlab libvirt config eos' in another
window and follow the instructions.

WARNING  --os-type is deprecated and does nothing. Please stop using it.
ERROR    
--os-variant/--osinfo OS name is required, but no value was
set or detected.

This is now a fatal error. Specifying an OS name is required
for modern, performant, and secure virtual machine defaults.

You can see a full list of possible OS name values with:

   virt-install --osinfo list

If your Linux distro is not listed, try one of generic values
such as: linux2020, linux2018, linux2016

If you just need to get the old behavior back, you can use:

  --osinfo detect=on,require=off

```

**Reply by ipspace** [reply]
> I built an Ubuntu 22.04.1 LTS VM on my Debian PC. Inside the VM, I was able to boot vEOS64-lab-4.28.2F.qcow2 using this page : https://codingpackets.com/blog/arista-veos-vagrant-libvirt-box-install/

Great, so it's some missing package on Debian. Unfortunately I don't know anything about Debian, so the only suggestion would be to go through the Ubuntu installation script (libvirt.sh and ubuntu.sh in https://github.com/ipspace/netlab/tree/dev/netsim/install) and see if something rings a bell

> FYI, the `netlab libvirt package eos vEOS64-lab-4.28.2F.qcow2` gives an error for deprecated option :

Thank you, will try to fix that.

**Reply by ipspace** [reply]
> I built an Ubuntu 22.04.1 LTS VM on my Debian PC. Inside the VM, I was able to boot vEOS64-lab-4.28.2F.qcow2 using this page : https://codingpackets.com/blog/arista-veos-vagrant-libvirt-box-install/ FYI, the `netlab libvirt package eos vEOS64-lab-4.28.2F.qcow2` gives an error for deprecated option

I'm guessing that the recipe from codingpackets gave you exactly the same error, as I copied the `virt-install` parameters from it (I'm not smart enough to figure them out on my own). Mentioning what parameters worked for you would have saved me quite some time 🤷‍♂️

Opened #485 to track this, any suggestions or patches that work on 20.04 and 22.04 are most welcome.

**Reply by ipspace** [reply]
#485 is fixed by 9fc2ebe15753f3cd83168d7f80092edfa4ca2575. You can download the source from GitHub or wait for the next release -- hope the fix will work for you.
